### ISSUE TITLE
[Feature Request] Stock management

### ISSUE THREAD

**Issue by BertBrug** [InvoicePlane/InvoicePlane-v2#406]
What i would like to see in the future is stock management. For example: I sell the following products:

- Cola
- Fanta
- Cassis
- Orange Juice

I would like to know how many of them are in stock so i can see what is going out and coming in to my shop.

**Comment by nielsdrost7** [comment]
Try to show your vision on how you would Ike to do that.
I click on.... I see ....
I sell ... and this happens to my stock for that product.
Things like that will help.

**Comment by dfear** [comment]
TBH, @BertBrug Invoiceplane is probably not the best / most appropriate software for your shop / business. What you need is inventory / POS focused software. 

Developing this feature request would pull core resources away from ACTUALLY releasing InvoicePlane version 2 some time soon. If you want to develop a stock control / inventory / POS addon, then go ahead and do so. All contributions are welcome.

**Comment by clockwiseq** [comment]
I'll throw my two cents in as well as I have worked with inventory systems in the past.  For us to implement inventory and stock levels, we will have to add a purchasing system.  We wouldn't have to necessarily implement a full-fledged request >> po >> purchase invoice model, but definitely at least something like a po or invoice would have to be created and it would include the items created in IP.  Once you create the PO, the stock level wouldn't increase, it would show that items have been ordered.  Once the items arrive, the purchasing clerk (or warehouse person) would mark the items that they received and the quantity.  That way, you have a reconciliation of what was ordered and what was received.  Once the items have been received, the PO is marked as received/processed.  A process would then kick off to update the current stock levels.  Something that other inventory systems do is they have a monthly job that runs to recalculate stock levels (recounts from the beginning of how many of the items you have, how many you have sold, and how many you have purchased.  We can write up some requirements for this.

**Comment by clockwiseq** [comment]
> TBH, @BertBrug Invoiceplane is probably not the best / most appropriate software for your shop / business. What you need is inventory / POS focused software.
> 
> Developing this feature request would pull core resources away from ACTUALLY releasing InvoicePlane version 2 some time soon. If you want to develop a stock control / inventory / POS addon, then go ahead and do so. All contributions are welcome.

I will agree with you there.  IP really is intended to be a simple invoicing platform.  BUT, if we can find spare resources to work on something like this, it wouldn't be a bad contribution.

**Comment by BertBrug** [comment]
Alright here is my explanation. For example i have a company that wants to monitor what kind of stock we have. For example:

When someone bought 250 cans of beans, i want them to be able to insert it in to IP. The reason for this is so i can see what whas gathered or what is the stock for a certain product i use.

I'm planning to use IP for a game related reason. So when my employees use IP i want them to insert the amount they have gathered. I would like to see the the following functions coming with this improvement:

- Add X amount of items to product stock
- Remove X amount of items to product stock

If u have any other questions please let me know so i can give u more details on it.

**Comment by nielsdrost7** [comment]
Thank you for the explanation.
If we would ever build it, we would build it according to what clockwiseq suggested.

Your explanation is great, but it's not sustainable.
It's for 1 shop and 1 shop only.

When we build something in IP, we have to keep it simple, buttttt... we have to build it with multiple possibilities in mind.

The simple part could still be what you're suggesting though.
And it will take a while before this feature request will be built.

**Comment by OPGRBILLINGDEV** [comment]
Suggestion is also that "this resource is unlimited" would be usefull  Saleor resource management for stock could be usefull. It's in the git, look it up.

**Comment by nielsdrost7** [comment]
## Planning Note — v2 Implementation

**Story:** As a company admin, I want to track stock quantities for products, so that I know when items are running low and can see how invoice line items affect available stock.

**v2 context:** v2 has a `Product` model in `Modules/Products/Models/Product.php`. The Products module is the right home for stock tracking. A minimal, v2-appropriate implementation adds a `stock_quantity` field to products and decrements it when an invoice containing that product is finalised — using the existing `InvoiceItem` model and Laravel events. Full warehouse management (locations, suppliers, purchase orders) is out of scope for an invoicing app and should be a separate future module.

**Acceptance Criteria:**
- Each `Product` has an optional `stock_quantity` (nullable integer; null = not tracked)
- Stock is decremented when an invoice transitions to status "Sent" or "Paid" (configurable)
- Stock is incremented (restored) when an invoice is voided or deleted
- The product list in Filament shows current stock with a visual low-stock warning below a configurable threshold
- A Filament action on the product allows manual stock adjustment with an audit note

**Implementation:**
- Migration: add `stock_quantity` (nullable unsignedInteger) and `stock_low_threshold` (nullable unsignedInteger, default 5) to `products` table in `Modules/Products/Database/Migrations/`
- Create `Modules/Invoices/Listeners/AdjustProductStock.php` — listens to `InvoiceStatusChanged` event, loops `$invoice->items`, finds matching `Product`, and decrements/increments `stock_quantity` using `DB::table('products')->decrement()`
- Register listener in `Modules/Invoices/Providers/InvoicesServiceProvider.php`
- Add `TextColumn::make('stock_quantity')->color(fn ($state, $record) => $state <= $record->stock_low_threshold ? 'danger' : 'success')` to `ProductResource` table
- Add `AdjustStockAction` (custom Filament `Action`) with a numeric input and reason textarea that calls `ProductService::adjustStock(Product $product, int $delta, string $reason)`; log adjustment to `AuditLog`

**Test Scenarios (PHPUnit):**
```php
class ProductStockTest extends AbstractCompanyPanelTestCase
{
    #[Test]
    public function stock_is_decremented_when_invoice_is_sent(): void
    {
        $product = Product::factory()->for($this->company)->create(['stock_quantity' => 10]);
        $invoice = Invoice::factory()->for($this->company)->create();
        InvoiceItem::factory()->for($invoice)->create(['product_id' => $product->id, 'quantity' => 3]);

        event(new InvoiceStatusChanged($invoice, InvoiceStatus::Sent));

        $this->assertDatabaseHas('products', ['id' => $product->id, 'stock_quantity' => 7]);
    }

    #[Test]
    public function stock_is_restored_when_invoice_is_voided(): void
    {
        $product = Product::factory()->for($this->company)->create(['stock_quantity' => 7]);
        $invoice = Invoice::factory()->for($this->company)->create(['status' => InvoiceStatus::Sent]);
        InvoiceItem::factory()->for($invoice)->create(['product_id' => $product->id, 'quantity' => 3]);

        event(new InvoiceStatusChanged($invoice, InvoiceStatus::Void));

        $this->assertDatabaseHas('products', ['id' => $product->id, 'stock_quantity' => 10]);
    }

    #[Test]
    public function null_stock_quantity_is_not_adjusted(): void
    {
        $product = Product::factory()->for($this->company)->create(['stock_quantity' => null]);
        $invoice = Invoice::factory()->for($this->company)->create();
        InvoiceItem::factory()->for($invoice)->create(['product_id' => $product->id, 'quantity' => 5]);

        event(new InvoiceStatusChanged($invoice, InvoiceStatus::Sent));

        $this->assertDatabaseHas('products', ['id' => $product->id, 'stock_quantity' => null]);
    }
}
```

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by nielsdrost7** [comment]
---

## Epic Status — Stock Management

_Added to support future implementation of this request. The issue body is unchanged._

> **Context/Epic**: Products
>
> This is an epic/tracking issue aggregating the stock-management sub-issues listed in its own checklist. This note tracks sub-issue completion rather than fabricating tests for the epic itself; each sub-issue gets its own plan when processed individually.

### Specific
Track delivery of end-to-end stock management: a `stock_quantity` field on products, real-time availability display, automatic deduction on invoice finalisation, quote-level stock consideration, and a central `StockAdjustmentService`. **Current state verified against the codebase**: none of this exists yet. `Modules/Products/Models/Product.php` and its migration (`2010_01_01_000019_create_products_table.php`) have no `stock_quantity` column; a repo-wide search for `stock` across `Modules/` returns zero matches outside this epic's own sub-issue write-ups. This is a fully greenfield epic, not a partially-built one.

### Measurable
- [ ] #21 — Add stock quantity field to products — not started; no `stock_quantity` column, form field, or table column exists. This is the foundational blocker for every other item below.
- [ ] #477 — Real-time availability display (Quantity column, low-stock badges, filters) — not started; depends on #21.
- [ ] #66 — Automatic stock deduction on invoice finalisation — not started; depends on #21. Note: this codebase has no explicit "finalise" action — `InvoiceStatus` is a plain Select field on the invoice form (`Modules/Invoices/Filament/Company/Resources/Invoices/Schemas/InvoiceForm.php`), set directly rather than via a dedicated workflow action, so "finalisation" needs to be defined as a status transition (e.g. any status change into `SENT`) before deduction logic can hook into it.
- [ ] #486 — Stock adjustment service (invoices, quotes, manual) — not started; no `StockAdjustmentService` class exists anywhere in `Modules/Products/Services/`. This is the epic's own acceptance criterion ("A `StockAdjustmentService` handles all programmatic adjustments") and should likely be built before #66/#68 rather than after, since both would otherwise duplicate deduction logic independently.
- [ ] #68 — Quote-level stock consideration — not started; depends on #21 and likely #486.
- [ ] #476 — Product attributes (related, tracked separately) — not started; see its own SMART Plan for a real gap analysis (a generic `CustomField`/`CustomFieldValue` system already exists company-wide but isn't wired to products).
- [ ] #478 — Profit margin (related, tracked separately) — partially groundable: `products.cost_price` already exists as a column, but `profit_margin_percent` and all margin-calculation/visibility logic do not.

**Definition of done for this epic**: all six numbered sub-issues above are closed, `Modules/Products/Tests/Feature/` covers stock behaviour end-to-end, and the epic's own acceptance criteria (company-scoped stock data, negative-stock guard, `StockAdjustmentService` as the single adjustment path) hold.

### Achievable
1. #21 (stock_quantity field) is the true root dependency — every other stock sub-issue is blocked on it and should not start first.
2. Build #486 (`StockAdjustmentService`) immediately after #21, before #66 and #68, so invoice-finalisation and quote-reservation deduction both route through one service rather than each invoice/quote implementing its own stock math (avoids the duplicate-logic trap the epic's own acceptance criteria are trying to prevent).
3. #477 (availability display) only needs #21, not #486 — it can be built in parallel with the service work since it's a read-only table/badge feature.
4. #476 (attributes) and #478 (margin) are independent of the stock-tracking chain entirely and can proceed in parallel with any of the above.

### Relevant
Stock tracking blocks over-selling/stockout detection, which is core inventory functionality for any user selling physical goods — but because nothing here exists yet, sequencing (field → service → consumers) matters more than usual to avoid rework.

### Time-bound
No fixed date — epic closes when its six sub-issues close. Re-review at each sprint boundary; #21 should be prioritized first as the blocking dependency for everything else in this epic.
